c++/src/kj/compat/http.c++

```diff
@@ -4746,6 +5271,153 @@ private:
     kj::Own<kj::PromiseFulfiller<HttpClient::WebSocketResponse>> fulfiller;
     kj::Promise<void> task = nullptr;
   };
+
+  class DelayedEofIoStream final: public kj::AsyncIoStream {
+    // An AsyncIoStream wrapper that, when it reaches EOF, delays the final read until some
+    // promise completes.
+  public:
+    DelayedEofIoStream(kj::Own<kj::AsyncIoStream> inner, kj::Promise<void> completionTask)
+        : inner(kj::mv(inner)), completionTask(kj::mv(completionTask)) {}
+
+    kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+      return wrap(minBytes, inner->tryRead(buffer, minBytes, maxBytes));
+    }
+
+    kj::Maybe<uint64_t> tryGetLength() override {
+      return inner->tryGetLength();
+    }
+
+    kj::Promise<uint64_t> pumpTo(kj::AsyncOutputStream& output, uint64_t amount) override {
+      return wrap(amount, inner->pumpTo(output, amount));
+    }
+
+    Promise<void> write(const void* buffer, size_t size) override {
+      return inner->write(buffer, size);
+    }
+
+    Promise<void> write(ArrayPtr<const ArrayPtr<const byte>> pieces) override {
+      return inner->write(pieces);
+    }
+
+    Promise<void> whenWriteDisconnected() override {
+      return inner->whenWriteDisconnected();
+    }
+
+    void shutdownWrite() override {
+      inner->shutdownWrite();
+    }
+
+  private:
+    kj::Own<kj::AsyncIoStream> inner;
+    kj::Maybe<kj::Promise<void>> completionTask;
+
+    template <typename T>
+    kj::Promise<T> wrap(T requested, kj::Promise<T> innerPromise) {
```

Comment:
I wonder if we could reuse `DelayedEofInputStream` rather than reproduce this logic. You could set up a `DelayedEofInputStream` wrapping the input side of `inner`, forward all read calls to that, while forwarding write calls directly to `inner`.